File: tangzer-front/src/types/ChatMessage.ts
export type MessageType = "CHAT" | "JOIN" | "LEAVE";

export type MessageDeliveryStatus = "SENT" | "DELIVERED" | "READ";

export interface ChatMessage {
    type: MessageType;
    content: string;
    sender: string;
    roomId: string | null;
    timestamp: string;
    messageId?: string;
    recipient?: string | null;
}

export interface MessageStatus {
    messageId: string;
    status: MessageDeliveryStatus;
    username: string;
}

export interface TypingStatus {
    username: string;
    recipient: string;
    typing: boolean;
}

File: src/main/java/net/ent/etnc/chattangzer/models/dtos/ChatMessage.java
package net.ent.etnc.chattangzer.models.dtos;

import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.Instant;

@AllArgsConstructor
@NoArgsConstructor
public class ChatMessage {

    @Getter @Setter
    private MessageType type;

    @Getter @Setter
    private String content;

    @Getter @Setter
    private String sender;

    @Getter @Setter
    private String roomId;

    @Getter @Setter
    private Instant timestamp;

    @Getter @Setter
    private String messageId;

    @Getter @Setter
    private String recipient;

    public ChatMessage(MessageType type, String content, String sender, String roomId) {
        this(type, content, sender, roomId, Instant.now(), null, null);
    }

    public ChatMessage(MessageType type, String content, String sender, String roomId, Instant timestamp) {
        this(type, content, sender, roomId, timestamp, null, null);
    }
}
